Bio/SeqIO/QualityIO.py

```diff
@@ -796,6 +801,115 @@ def _get_solexa_quality_str(record):
+        are checked here.  The first line of the record must begin with a '@'.
+        The second line, which is the read sequence, must not have any white
+        space within the sequence. Next, the length of the read sequence and
+        the quality sequence must be equal.
+        """
+        # Record level check 1: Check if start of title line is @
+        if line_1[0] != "@":
+            raise ValueError("FASTQ records should start with '@' character.")
+        # Record level check 2: Check if start of second title line is +
+        if line_3[0] != "+":
+            raise ValueError("Quality title should start with '+' character.")
+        # Record level check 3: If quality caption exists, check for equality
+        if line_3[1:-1] and line_3[1:] != line_1[1:]:
+            raise ValueError("Sequence and quality captions differ.")
+        # Record level check 4: Ensure there is no white space in the sequence
+        if " " in line_2 or "\t" in line_2:
+            raise ValueError("Whitespace is not allowed in the sequence.")
+        # Record level check 5: Ensure the sequence and quality are of same length
+        if len(line_2) != len(line_4):
+            raise ValueError("Lengths of sequence and quality values differ "
+                             " for %s (%i and %i)."
+                             % (line_1[1:-1], len(line_2), len(line_4)))
+
+    # File level check 1: Ensure the file is not empty
+    try:
+        first_1 = next(handle)
+    except StopIteration as e:
+        return
+    # File level check 2: check if file is in binary format
+    if first_1[0].isdigit():
+        raise ValueError("Is this handle in binary mode not text mode?")
+
+    # Record level check 0: Ensure the first record is complete
+    try:
+        first_2 = next(handle).rstrip()
+        first_3 = next(handle)
+        first_4 = next(handle).rstrip()
+    except StopIteration as e:
+        raise suppress_context(ValueError('Incomplete record encountered'))
+    # Record level checks 1, 2, 3, 4, 5: Defined within check()
+    check(first_1, first_2, first_3, first_4)
+    yield (first_1[1:-1], first_2, first_4)
+
+    # Record level check 0: Ensure the record is complete
+    try:
+        # zip the iterators, one for each line, and iterate over them
+        for line_1, line_2, line_3, line_4 in zip_longest(
+                handle, handle, handle, handle):
+            # Avoid stripping newlines from line 1 and line 3 for speed
+            line_2, line_4 = line_2.rstrip(), line_4.rstrip()
+            # Record level checks 1, 2, 3, 4, 5: Defined within check()
+            check(line_1, line_2, line_3, line_4)
+            yield (line_1[1:-1], line_2, line_4)
+    except AttributeError as e:
+        # If a record is incomplete, zip_longest uses None as the default
+        # fillvalue instead, which raises an AttributeError on using rstrip()
+        raise suppress_context(ValueError('Incomplete record encountered'))
```

Comment:
consider just checking `if line_4 is None` at the end instead (if it's ok to `yield` the incomplete record before raising the error?). If you keep the try-except style then you should `try` just the line with `rstrip` calls on it, rather than the whole code block

Reply:
I think it is more Pythonic (also slightly faster) to use try-except than use the if block (EAFP) but yes, shifting the checking to just the rstrip statement makes sense. Thanks!